### ISSUE TITLE
[0.9.2] initscript restart doesn't wait for process to shutdown

### ISSUE THREAD

**Issue by johnl** [influxdb/influxdb#3548]
Installed InfluxDB v0.9.2 (git: 6c0a91f775f9fc0e625d17ffa04a3fe86945ba09) from the deb package onto 64bit Ubuntu Trusty, enabled the graphite and collectd listeners and I've got a few gig of datapoints.

When I use the initscript to restart influxdb, it doesn't seem to wait for the existing process to finish shutting down and I end up with two processes running for a while, until the first finally finished shutting down.

```
root@srv-xtttc:/var/opt/influxdb# service influxdb restart
influxd process was stopped [ OK ]
Starting the process influxd [ OK ]
influxd process failed to start [ FAILED ]

root@srv-xtttc:/var/opt/influxdb# ps auxf | grep influxd

influxdb  9572 15.9 15.7 917376 322996 pts/0   Sl   15:27  27:37 /opt/influxdb/influxd -pidfile /var/run/influxdb/influxd.pid -config /etc/opt/influxdb/influxdb.conf
influxdb  9815 64.7 15.8 6752260 325024 pts/0  Rl   18:20   0:05 /opt/influxdb/influxd -pidfile /var/run/influxdb/influxd.pid -config /etc/opt/influxdb/influxdb.conf


root@srv-xtttc:/var/opt/influxdb# tail /var/log/influxdb/influxd.log
[metastore] 2015/08/03 18:20:25 [DEBUG] raft: Vote granted. Tally: 1
[metastore] 2015/08/03 18:20:25 [INFO] raft: Election won. Tally: 1
[metastore] 2015/08/03 18:20:25 [INFO] raft: Node at 127.0.0.1:8088 [Leader] entering Leader state
[metastore] 2015/08/03 18:20:25 [INFO] raft: Disabling EnableSingleNode (bootstrap)
[metastore] 2015/08/03 18:20:25 [DEBUG] raft: Node 127.0.0.1:8088 updated peer set (2): [127.0.0.1:8088]
[metastore] 2015/08/03 18:20:25 [DEBUG] raft: Node 127.0.0.1:8088 updated peer set (2): [127.0.0.1:8088]
[metastore] 2015/08/03 18:20:25 [DEBUG] raft: Node 127.0.0.1:8088 updated peer set (2): [127.0.0.1:8088]
[metastore] 2015/08/03 18:20:25 [DEBUG] raft: Node 127.0.0.1:8088 updated peer set (2): [127.0.0.1:8088]
[metastore] 2015/08/03 18:20:25 [DEBUG] raft: Node 127.0.0.1:8088 updated peer set (2): [127.0.0.1:8088]
[shard] 2015/08/03 18:20:26 flush 19152 points in 1.197s
[shard] 2015/08/03 18:20:27 flush 19735 points in 1.039s
[shard] 2015/08/03 18:20:28 flush 18287 points in 1.045s
[shard] 2015/08/03 18:20:29 flush 17815 points in 0.991s
[shard] 2015/08/03 18:20:30 flush 19025 points in 0.943s
[shard] 2015/08/03 18:20:31 flush 17841 points in 0.984s
[run] 2015/08/03 18:20:51 time limit reached, initializing hard shutdown
```

**Comment by johnl** [comment]
the initscript also reports that the start failed, despite it succeeding.

**Comment by tuukkamustonen** [comment]
I have something similar with `0.9.2` also. Running debian wheezy, `sudo service influxdb stop`:

    [run] 2015/08/04 13:50:45 waiting for clean shutdown...
    [snapshot] 2015/08/04 13:50:45 snapshot listener closed
    [tcp] 2015/08/04 13:50:45 cluster service accept error: network connection closed
    [shard-precreation] 2015/08/04 13:50:45 precreation service terminating
    [retention] 2015/08/04 13:50:45 retention policy enforcement terminating
    [retention] 2015/08/04 13:50:45 retention policy enforcement terminating

Wait for 20-30 seconds or so and:

    [run] 2015/08/04 13:51:15 time limit reached, initializing hard shutdown

After this, the process is actually terminated and `service influxdb start` works.

`sudo service influxdb restart` gives:

    [run] 2015/08/04 13:53:24 signal received, initializing clean shutdown...
    [run] 2015/08/04 13:53:24 waiting for clean shutdown...
    [tcp] 2015/08/04 13:53:24 cluster service accept error: network connection closed
    [snapshot] 2015/08/04 13:53:24 snapshot listener closed
    [shard-precreation] 2015/08/04 13:53:24 precreation service terminating
    [retention] 2015/08/04 13:53:24 retention policy enforcement terminating
    [retention] 2015/08/04 13:53:24 retention policy enforcement terminating
    [metastore] 2015/08/04 13:53:24 [INFO] raft: Node at 127.0.0.1:8088 [Follower] entering Follower state
    [metastore] 2015/08/04 13:53:24 read local node id: 1
    [shard] 2015/08/04 13:53:24 flush 2 points in 0.002s
    [shard] 2015/08/04 13:53:24 flush 4 points in 0.002s
    [shard] 2015/08/04 13:53:24 flush 3 points in 0.002s
    [shard] 2015/08/04 13:53:24 flush 3 points in 0.002s
    [admin] 2015/08/04 13:53:24 listening on HTTP: [::]:8083
    [httpd] 2015/08/04 13:53:24 authentication enabled: false
    [httpd] 2015/08/04 13:53:24 listening on HTTP: [::]:8086
    Job for influxdb.service failed. See 'systemctl status influxdb.service' and 'journalctl -xn' for details.
    [metastore] 2015/08/04 13:53:26 [WARN] raft: Heartbeat timeout reached, starting election
    [metastore] 2015/08/04 13:53:26 [INFO] raft: Node at 127.0.0.1:8088 [Candidate] entering Candidate state
    [metastore] 2015/08/04 13:53:26 [DEBUG] raft: Votes needed: 1
    [metastore] 2015/08/04 13:53:26 [DEBUG] raft: Vote granted. Tally: 1
    [metastore] 2015/08/04 13:53:26 [INFO] raft: Election won. Tally: 1
    [metastore] 2015/08/04 13:53:26 [INFO] raft: Node at 127.0.0.1:8088 [Leader] entering Leader state
    [metastore] 2015/08/04 13:53:26 [INFO] raft: Disabling EnableSingleNode (bootstrap)
    [metastore] 2015/08/04 13:53:26 [DEBUG] raft: Node 127.0.0.1:8088 updated peer set (2): [127.0.0.1:8088]
    [metastore] 2015/08/04 13:53:26 [DEBUG] raft: Node 127.0.0.1:8088 updated peer set (2): [127.0.0.1:8088]
    [metastore] 2015/08/04 13:53:26 [DEBUG] raft: Node 127.0.0.1:8088 updated peer set (2): [127.0.0.1:8088]
    [metastore] 2015/08/04 13:53:26 [DEBUG] raft: Node 127.0.0.1:8088 updated peer set (2): [127.0.0.1:8088]
    [metastore] 2015/08/04 13:53:26 [DEBUG] raft: Node 127.0.0.1:8088 updated peer set (2): [127.0.0.1:8088]
    [metastore] 2015/08/04 13:53:26 [DEBUG] raft: Node 127.0.0.1:8088 updated peer set (2): [127.0.0.1:8088]
    [metastore] 2015/08/04 13:53:26 [DEBUG] raft: Node 127.0.0.1:8088 updated peer set (2): [127.0.0.1:8088]
    [metastore] 2015/08/04 13:53:26 [DEBUG] raft: Node 127.0.0.1:8088 updated peer set (2): [127.0.0.1:8088]
    [graphite] 2015/08/04 13:53:26 ensured target database graphite exists
    [tcp] 2015/08/04 13:53:26 cluster service accept error: network connection closed
    [snapshot] 2015/08/04 13:53:26 snapshot listener closed
    [shard-precreation] 2015/08/04 13:53:26 precreation service terminating
    [retention] 2015/08/04 13:53:26 retention policy enforcement terminating
    [retention] 2015/08/04 13:53:26 retention policy enforcement terminating
    run: open server: open service: listen tcp :2003: bind: address already in use

Then wait again 20-30 seconds and:

    [run] 2015/08/04 13:53:54 time limit reached, initializing hard shutdown

This happens only with graphite plugin enabled. Seems like it does not want to shutdown? Do you have graphite plugin enabled @johnl ?

**Comment by johnl** [comment]
yeah, I mentioned that. graphite and collectd listeners enabled - the only changes to the default config I think.

but that might be a separate issue - the initscripts shouldn't consider the daemon shut down if the process hasn't actually finished writing to disk, no matter what caused it to hang on :)

**Comment by tuukkamustonen** [comment]
@johnl Indeed you did, no idea how I was so blind :O

True, my problem is probably different, but I would assume they are related. Let's see if I need to spin it off to a separate ticket.

**Comment by poll0rz** [comment]
+1, same things happening here

**Comment by otoolep** [comment]
Fixed by https://github.com/influxdb/influxdb/pull/3980

**Comment by otoolep** [comment]
All fixed for 0.9.4.
